File: c/day2/day2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <memory.h>

int main() {
    FILE *fp = fopen("input.txt", "r");
    if (fp == NULL) {
        return EXIT_FAILURE;
    }
    int c = 0;
    int ln = 0;
    int col = 0;
    char line[100];
    int safe = 0;
    while (1) {
        c = fgetc(fp);
        if (c == '\n' || c == EOF) {
            col = 0;
            char *tok = strtok(line, " ");
            char *ntok = strtok(NULL, " ");
            int *diffs = malloc(0);
            int i = 0;
            while (ntok != NULL) {
                int a = atoi(tok);
                int b = atoi(ntok);
                int dif = b - a;
                diffs = realloc(diffs, sizeof(int) * (i+1));
                diffs[i] = dif;
                strcpy(tok, ntok);
                ntok = strtok(NULL, " ");
                i++;
            }
            int ok = 1;
            for (int j = 0; j < i; j++) {
                if (abs(diffs[j]) < 1 || abs(diffs[j]) > 3) {
                    ok = 0;
                    break;
                }
            }

            int dec = 1;
            for (int j = 0; j < i; j++) {
                if (diffs[j] > 0) {
                    dec = 0;
                    break;
                }
            }

            int inc = 1;
            for (int j = 0; j < i; j++) {
                if (diffs[j] < 0) {
                    inc = 0;
                    break;
                }
            }

            if (!dec && !inc) {
                ok = 0;
            }
            if (dec && inc) {
                ok = 0;
            }

            safe += ok;
            ln++;
            if (c == EOF) {
                break;
            }
        } else {
            line[col] = c;
            col++;
        }
    }
    printf("%d\n", safe);
    fclose(fp);
}

File: py/day2/day2.py
if __name__ == "__main__":
    with open("input.txt") as fp:
        safe = 0
        for record in fp.readlines():
            levels = list(map(lambda x: int(x), record.split(" ")))
            ok = []
            for j in range(len(levels)):
                levelsafe = True
                rem = levels[:j] + levels[j+1:]
                diffs = [rem[i+1]-rem[i] for i in range(len(rem)-1)]
                if any(abs(d) < 1 or abs(d) > 3 for d in diffs):
                    levelsafe = False
                inc = all(d < 0 for d in diffs)
                dec = all(d > 0 for d in diffs)
                if not inc and not dec:
                    levelsafe = False
                if inc and dec:
                    levelsafe = False
                ok.append(levelsafe)
            if any(ok):
                safe += 1
        print(safe)
            
